### ISSUE TITLE
Question on component detection

### ISSUE THREAD

**Issue by hriss95** [yannickcr/eslint-plugin-react#867]
Hello, I've been going through the tests for the `displayName` rule and I could not figure out how component detection exactly works (consulted with the code) after seeing one test. In particular, I am talking about:

```
var Mixins = {
  Greetings: {
    Hello: function() {
      return <div>Hello {this.props.name}</div>;
    }
  }
}
Mixins.Greetings.Hello.displayName = 'Hello';
```

How is the `Hello` property in an object literal different from a `render` property in a component created using the ES5 method? Looking at the code, I thought these are cases in which the component might be added from the `ReturnStatement` detection instruction, but I struggle to see how the two cases(`render` in ES5 components and `Hello` in this example) are different.

**Comment by ljharb** [comment]
Your component here is an SFC that happens to be nested under two objects - something with `render` is likely to be a class-based or createClass component.

Is the problem that you're code example is being detected as a component, or that it is not?

**Comment by hriss95** [comment]
My question is regarding more of the implementation level details for the plug-in. This code example is in the tests for the `display-name.js` rule in the plug-in and claims that `Hello` is a component. My question is how that is detected since this is syntactically the same to a `render` method in ES5 components. For example, the tests contain things like:
```
var Hello = React.createClass({
  displayName: 'Hello',
  render: function() {
    return <div>Hello {this.props.name}</div>;
  }
});
```

`render` here is syntactically the same as it is a property in an object literal which is initiated with a function expression and the function returns JSX content. However, it would not be detected as a component. So, how does the plug-in distinguish between the two cases?

**Comment by ljharb** [comment]
The word "render" is special, and a function with a capital latter name in an object literal would be a stateless functional components.

**Comment by hriss95** [comment]
So, just to clarify: I know render is special, but is it not a component because its special or because its a method with an identified that starts with a lower case letter?

**Comment by ljharb** [comment]
Lowercase names are for DOM components in React. Component names only start with an uppercase letter.
